Update typing for Python 3.7 (1)

cdce8p opened the pull request:
## Description
Part 1 of 3 to update the typing in astroid following the bump to Python 3.7.2

Part 1: Replace `typing` aliases for builtin types
Part 2: Replace `typing` aliases for `collections.abc` classes.
Part 3: Update `py-version` in `pylintrc`

DanielNoord commented on astroid/brain/brain_builtin_inference.py:
```diff
@@ -3,10 +3,11 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Astroid hooks for various builtins."""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/brain/brain_namedtuple_enum.py:
```diff
@@ -3,11 +3,12 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Astroid hooks for the Python standard library."""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/filter_statements.py:
```diff
@@ -5,15 +5,14 @@
 """
-
-from typing import List, Optional, Tuple
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/inference.py:
```diff
@@ -4,24 +4,13 @@
 
 """this module contains a set of functions to handle inference on astroid trees
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/inference_tip.py:
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Transform utilities (filters and decorator)"""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/interpreter/objectmodel.py:
```diff
@@ -20,13 +20,14 @@
 attribute. Thus the model can be viewed as a special part of the lookup
 mechanism.
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/manager.py:
```diff
@@ -6,12 +6,13 @@
 possible by providing a class responsible to get astroid representation
 from various source and using a cache of built modules)
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/mixins.py:
```diff
@@ -4,9 +4,11 @@
 
 """This module contains some mixins for the different nodes.
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/modutils.py:
```diff
@@ -13,6 +13,7 @@
 :type BUILTIN_MODULES: dict
 :var BUILTIN_MODULES: dictionary with builtin module names has key
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/nodes/node_classes.py:
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Module for some node classes. More nodes in scoped_nodes.py"""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/nodes/scoped_nodes/mixin.py:
```diff
@@ -3,8 +3,9 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """This module contains mixin classes for scoped nodes."""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/nodes/scoped_nodes/scoped_nodes.py:
```diff
@@ -7,23 +7,15 @@
 new local scope in the language definition : Module, ClassDef, FunctionDef (and
 Lambda, GeneratorExp, DictComp and SetComp to some extent).
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/nodes/scoped_nodes/utils.py:
```diff
@@ -5,20 +5,21 @@
 """
 This module contains utility functions for scoped nodes.
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/objects.py:
```diff
@@ -10,9 +10,10 @@
 
     Call(func=Name('frozenset'), args=Tuple(...))
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/protocols.py:
```diff
@@ -5,11 +5,12 @@
 """this module contains a set of functions to handle python protocols for nodes
 where it makes sense.
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/rebuilder.py:
```diff
@@ -5,26 +5,14 @@
 """this module contains utilities for rebuilding an _ast tree in
 order to get a single Astroid representation
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on astroid/test_utils.py:
```diff
@@ -3,11 +3,13 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Utility functions for test code that uses astroid ASTs as input."""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on pylintrc:
```diff
@@ -310,7 +310,7 @@ mixin-class-rgx=.*Mix[Ii]n
 # system, and so shouldn't trigger E0201 when accessed. Python regular
 # expressions are accepted.
-generated-members=REQUEST,acl_users,aq_parent,argparse.Namespace
+generated-members=REQUEST,acl_users,aq_parent,argparse.Namespace,ast\.([mM]atch.*|pattern)
```
? Shouldn't this be in a different PR/commit?

cdce8p replied:
Need this to fix a `no-member` issue now that the string annotations are replaced with `from __future__ import annotations`, see https://github.com/PyCQA/pylint/issues/6594.

DanielNoord replied:
Ah! That explains it!

DanielNoord commented on script/bump_changelog.py:
```diff
@@ -6,12 +6,13 @@
 This script permits to upgrade the changelog in astroid or pylint when releasing a version.
 """
 # pylint: disable=logging-fstring-interpolation
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on tests/unittest_brain.py:
```diff
@@ -3,12 +3,14 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Tests for basic functionality in astroid.brain."""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on tests/unittest_inference.py:
```diff
@@ -3,13 +3,14 @@
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
 """Tests for the astroid inference capabilities"""
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord commented on tests/unittest_nodes.py:
```diff
@@ -4,12 +4,14 @@
 
 """tests for specific behaviour of astroid nodes
 """
+from __future__ import annotations
```
```suggestion

from __future__ import annotations
```

DanielNoord reviewed (commented):
I added a white line between imports and docstring whenever there originally was one.

Another approach would be to `replace all` on the `__future__` imports with a white lines in front of it. `Black` should then remove those case where there are more than one line and we would have consitency across all files.

Rest LGTM!

cdce8p commented:
> I added a white line between imports and docstring whenever there originally was one.

Added empty lines between every docstring and future import (in the changed files). There is an `isort` setting [`lines-before-imports`](https://pycqa.github.io/isort/docs/configuration/options.html#lines-before-imports). Unfortunately though it doesn't seem to work reliably for our use case. More promising is a black PR to make that the default behavior: https://github.com/psf/black/pull/2996.
